### PR TITLE
Add support for Auth Emulator

### PR DESCRIPTION
**Pull request by samtstern**
See #1849 

### Checklist:

- [x] Add useEmulator method to AuthUI
- [x] Centralize FirebaseAuth Instances
- [x] Add useEmulator to sample app
- [x] Make sample app choice permanent
- [x] Make all IDPs go through generic IDP while emulated
- [x] Changelog

### Manual Testing Completed:

- [x] Google sign in without emulator (just to be sure)
- [x] Google sign in with emulator
- [x] Google sign in with emulator, no other providers enabled (skip choice screen)
- [x] Email/password sign in with emulator
- [x] Yahoo sign in with emulator
- [x] Phone auth with emulator
- [x] Email --> welcome back Google with emulator

### Manual Testing Issues

- [ ] Google --> welcome back Email with emulator

This flow is not possible to to test, here's how the sequence goes:

1. Create email password user foo@bar.com
2. Attempt Google Sign In with emulator
3. Click "add new user" and create a google user with foo@bar.com email
4. Rather than detecting a collision, the Google user  seems to wipe out the email one every time

### CODE REVIEW

File: auth/src/main/java/com/firebase/ui/auth/viewmodel/email/WelcomeBackPasswordHandler.java [hunk 1]
```diff
@@ -71,6 +71,7 @@ public void startSignIn(@NonNull final String email,
         if (authOperationManager.canUpgradeAnonymous(getAuth(), getArguments())) {
             final AuthCredential credToValidate = EmailAuthProvider.getCredential(email, password);
 
+            // TODO: Audit this use of SOCIAL_PROVIDERS
```
Comment (samtstern) [inline auth/src/main/java/com/firebase/ui/auth/viewmodel/email/WelcomeBackPasswordHandler.java:74]: TODO(me): Do this

File: auth/src/main/java/com/firebase/ui/auth/ui/idp/SingleSignInActivity.java [hunk 2]
```diff
@@ -90,6 +93,7 @@ protected void onCreate(@Nullable Bundle savedInstanceState) {
         mProvider.getOperation().observe(this, new ResourceObserver<IdpResponse>(this) {
             @Override
             protected void onSuccess(@NonNull IdpResponse response) {
+                // TODO: Audit this use of SOCIAL_PROVIDERS
```
Comment (samtstern) [inline auth/src/main/java/com/firebase/ui/auth/ui/idp/SingleSignInActivity.java:96]: TODO(me): Do this

Comment (malcolmdeck) [reply auth/src/main/java/com/firebase/ui/auth/ui/idp/SingleSignInActivity.java:96]: Bump - not sure what this means, but apparently you should do it :P

File: auth/src/main/java/com/firebase/ui/auth/viewmodel/email/WelcomeBackPasswordHandler.java (hunk 1 above)
Comment (malcolmdeck) [reply auth/src/main/java/com/firebase/ui/auth/viewmodel/email/WelcomeBackPasswordHandler.java:74]: +1

File: auth/src/test/java/com/firebase/ui/auth/viewmodel/GenericIdpAnonymousUpgradeLinkingHandlerTest.java [hunk 3]
```diff
@@ -152,4 +152,13 @@ private void setupAnonymousUpgrade() {
         when(mMockUser.isAnonymous()).thenReturn(true);
         when(mMockAuth.getCurrentUser()).thenReturn(mMockUser);
     }
+
+    private void mockOAuthProvider(String providerId) {
+        // TODO(samstern): I wish we did not have to do this but the OAuthProvider() builder
+        //                 throws a NPE and we can't fix it due to b/172544960
```
Comment (malcolmdeck) [inline auth/src/test/java/com/firebase/ui/auth/viewmodel/GenericIdpAnonymousUpgradeLinkingHandlerTest.java:158]: I'll get to that later today, sorry it's causing you trouble

**Review by malcolmdeck — APPROVED** [review]
LGTM - this doesn't do anything I wouldn't expect it to, and the refactor for cleanliness seems appropriate. Thanks!

File: auth/src/main/java/com/firebase/ui/auth/data/remote/GenericIdpSignInHandler.java [hunk 4]
```diff
@@ -219,19 +241,29 @@ protected OAuthProvider buildOAuthProvider(String providerId) {
...
-    protected void handleSuccess(@NonNull String providerId,
+    protected void handleSuccess(boolean isUseEmulator,
+                                 @NonNull String providerId,
                                  @NonNull FirebaseUser user,
                                  @NonNull OAuthCredential credential,
                                  boolean isNewUser,
                                  boolean setPendingCredential) {
+
+        String accessToken = isUseEmulator
+                ? "fake_access_token"
```
[the remark this answers is not in the document]
Comment (samtstern) [reply auth/src/main/java/com/firebase/ui/auth/data/remote/GenericIdpSignInHandler.java:252]: Yeah it doesn't, but I'll change the code so if it does we use them.

File: auth/src/test/java/com/firebase/ui/auth/viewmodel/GenericIdpAnonymousUpgradeLinkingHandlerTest.java (hunk 3 above)
Comment (samtstern) [reply auth/src/test/java/com/firebase/ui/auth/viewmodel/GenericIdpAnonymousUpgradeLinkingHandlerTest.java:158]: Thanks for getting to this!

**Pull request merged**